crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -219,13 +269,168 @@ impl GraphQLClient {
+        let message = [
+            &response["message"],
+            &response["error"]["message"],
+            &response["errors"][0]["message"],
+        ]
+        .iter()
+        .map(|x| x.as_str())
+        .find(Option::is_some)
+        .flatten()
+        .unwrap_or("No message provided")
+        .to_string();
+
+        return Err(Error::InvalidReqwestStatus { status, message });
     }
+    Ok(())
 }
 
-#[allow(clippy::needless_pass_by_value)]
-fn to_execution_error(e: Error) -> DataFusionError {
-    DataFusionError::Execution(format!("{e}"))
+fn handle_graphql_query_error(response: &Value, query: &str) -> Result<()> {
+    let graphql_error = &response["errors"][0];
+    if !graphql_error.is_null() {
+        let line = usize::try_from(graphql_error["locations"][0]["line"].as_u64().unwrap_or(0))
+            .unwrap_or_default();
+        let column = usize::try_from(
+            graphql_error["locations"][0]["column"]
+                .as_u64()
+                .unwrap_or(0),
+        )
+        .unwrap_or_default();
+        return Err(Error::InvalidGraphQLQuery {
+            message: graphql_error["message"]
+                .as_str()
+                .unwrap_or_default()
+                .split(" at [")
+                .next()
+                .unwrap_or_default()
+                .to_string(),
+            line,
+            column,
+            query: format_query_with_context(query, line, column),
+        });
+    }
+    Ok(())
+}
+
+fn format_query_with_context(query: &str, line: usize, column: usize) -> String {
+    let query_lines: Vec<&str> = query.split('\n').collect();
+    let error_line = query_lines.get(line - 1).unwrap_or(&"");
+    let marker = " ".repeat(column - 1) + "^";
+    if line > 1 {
+        format!(
+            "{:>4} | {}\n{:>4} | {}\n{:>4} | {}",
+            line - 1,
+            query_lines[line - 2],
+            line,
+            error_line,
+            "",
+            marker
+        )
+    } else {
+        format!("{:>4} | {}\n{:>4} | {}", line, error_line, "", marker)
+    }
+}
+
+struct GraphQLTableProvider {
+    client: GraphQLClient,
+    schema: SchemaRef,
+}
+
+impl GraphQLTableProvider {
+    async fn new(client: GraphQLClient) -> Result<Self> {
+        let (_, schema, _) = client.execute(None, None, None).await?;
+
+        Ok(Self { client, schema })
+    }
+}
+
+#[async_trait]
+impl TableProvider for GraphQLTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        let res = self
+            .client
+            .execute_paginated(Arc::clone(&self.schema), limit)
```

Comment:
Please double check it is safe to pass limit here w/o filter? Let's say "select users where name == 'Kirill' limit 1". Does execute return first/single person due to limit 1 and then apply where after that?

Reply:
@sgrebnov that is good catch

Reply:
Added a test for this